### PR TITLE
build with dotnet 8 sdk

### PR DESCRIPTION
**Pull request by bruno-garcia**
When trying to run tests on .NET 8 (https://github.com/getsentry/sentry-dotnet/pull/2617/) it was clear that it was too much of a jump. Mobile support issues came up (Android and iOS, different issues). .NET 6 Android support was completely removed from .NET 8, so we need to chose between the two: https://github.com/xamarin/xamarin-android/pull/7900

This PR attempts to do the minimum changes required to build the repo with the .NET 8 SDK

This can be reverted possibly (as fixed in https://github.com/xamarin/xamarin-macios/pull/18793)
* https://github.com/getsentry/sentry-dotnet/blob/2f11b3073e4169a60cc8143b8d5bd6016a3b826a/samples/Directory.Build.props#L38

## Depends on:
* [x] https://github.com/getsentry/sentry-dotnet/pull/2624
* ~https://github.com/dotnet/sdk/issues/34496#issuecomment-1721875830~
* [x] https://github.com/getsentry/sentry-dotnet/issues/2639
* [x] https://github.com/xamarin/xamarin-macios/pull/18793 (should be part of .NET 8 RC2)

Since there's no customer-facing change:

#skip-changelog

### CODE REVIEW

**Comment by bruno-garcia** [comment]
CI fails due to: https://github.com/getsentry/sentry-dotnet/issues/2639

**Review by vaind — APPROVED** [review]

File: .github/workflows/build.yml
Comment (jamescrosswell) [inline .github/workflows/build.yml:18]: Pinned? Maybe put a comment in there explaining why so we know in the future whether this is still required or not.

File: .github/actions/environment/action.yml
Comment (jamescrosswell) [inline .github/actions/environment/action.yml:55]: Can we include an explanation for this? We just ran `dotnet workload restore` right? Won't doing `dotnet workload update` give us non-deterministic builds (we don't know what version of the workloads is going to be used for the build)?

**Comment by vaind** [comment]
After talking to @bitsandfoxes, I'm merging this to unblock other PRs and he'll follow up the review comments on `feat/4.0.0` branch

**Comment by bruno-garcia** [comment]
Thanks!

**Pull request merged**
